iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Shared;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace SimpleThermostat
+{
+    internal enum StatusCode
+    {
+        Completed = 200,
+        InProgress = 202,
+    }
+
+    public class Program
+    {
+        private const string ModelId = "dtmi:com:example:Thermostat;1";
+
+        private static readonly string s_deviceConnectionString = Environment.GetEnvironmentVariable("IOTHUB_DEVICE_CONNECTION_STRING");
+        private static readonly Random s_random = new Random();
+
+        private static DeviceClient s_deviceClient;
+
+        private static double s_temperature = 0d;
+
+        // Dictionary to hold the temperature updates sent over.
+        private static readonly Dictionary<DateTimeOffset, double> s_temperatureReadings = new Dictionary<DateTimeOffset, double>();
+
+        public static async Task Main(string[] _)
+        {
+            await RunSampleAsync();
+        }
+
+        private static async Task RunSampleAsync()
+        {
+            // This sample follows the following workflow:
+            // -> Initialize device client instance.
+            // -> Set handler to receive "targetTemperature" updates, and send the received update over reported property.
+            // -> Set handler to receive "getMaxMinReport" command, and send the generated report as command response.
+            // -> Periodically send "current temperature" over telemetry and "max temperature since last reboot" over property update.
+
+            PrintLog($"Initialize the device client.");
+            await InitializeDeviceClientAsync();
+
+            PrintLog($"Set handler to receive \"target temperature\" updates.");
+            await s_deviceClient.SetDesiredPropertyUpdateCallbackAsync(TargetTemperatureUpdateCallbackAsync, s_deviceClient);
+
+            PrintLog($"Set handler for \"getMaxMinReport\" command");
+            await s_deviceClient.SetMethodHandlerAsync("getMaxMinReport", HandleMaxMinReportCommandAsync, s_deviceClient);
+
+            bool temperatureReset = true;
+            await Task.Run(async () =>
+            {
+                while (true)
+                {
+                    if (temperatureReset)
+                    {
+                        // Generate a random value between 5.0°C and 45.0°C for the current temperature reading.
+                        s_temperature = Math.Round(s_random.NextDouble() * 40.0 + 5.0, 1);
+                    }
+
+                    await SendTemperatureTelemetryAsync();
+                    await SendMaxTemperatureSinceLastRebootAsync();
```

Comment:
there is no "reboot" command in `Thermostat` model, so this property seems weirdly placed. There is an ongoing discussion with Eric on this.
@bikamani , @vinagesh  FYI.